working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+          book.description = request_body["description"]
+      except:
+          abort(make_response({"message":"description is required"}, 400))
+
+      db.session.commit()
+
+      return make_response(jsonify(f"Book #{book.id} successfully updated"))
+  ```
+
+  **Updated tests:**
+  ```
+  def test_update_book_without_title(client, two_saved_books):
+      # Arrange
+      book_id = 1
+      data = {
+          "description": "New description"
+      }
+
+      # Act
+      response = client.put(f"/books/{book_id}", json=data)
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message": "title is required"}
+
+  def test_update_book_without_description(client, two_saved_books):
+      # Arrange
+      book_id = 1
+      data = {
+          "title": "Updated Book"
+      }
+
+      # Act
+      response = client.put(f"/books/{book_id}", json=data)
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message": "description is required"}
+  ```
+</details>
+
+Now that we've seen several facets of working with Copilot to write tests, we will leave it as a personal exercise to complete the test suite for our last untested dependency `delete_book`.
+
+#### Executing the Refactor
+
+Now that we have a solid test suite, we can get some help from Copilot to make this refactor go a little quicker. Let's review what needs to change in `validate_book` to make the function more flexible so we can use it with other models:
+- take in a second parameter that represents a model class reference
+- change any uses of the `Book` class to use our new parameter
+- update the function name, variable names, and return messages so they reflect that the function is not specific to the `Book` class
+
+The changes we need are pretty minimal and could be quick to perform manually, but for the practice, let's see how Copilot handles the change. If we highlight the `validate_book` function, open up the inline Copilot chat, and use the prompt:
+
+> Please update this function so that we could use it for any model class
+
+The response delivers exactly what we're looking for:
+```
+def validate_model(model_class, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} invalid"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} not found"}, 404))
+
+    return model
+```
+
+Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We can do this manually using VS Code's Find & Replace tools then adding in our new parameter, or we can reach out to Copilot. 
```

Comment:
- added additional change to name the clas reference parameter `cls`
```suggestion
Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We may also decide to rename `model_class` to the more common `cls`, which is often used in Python when passing in a class reference. We can do all of this manually using VS Code's Find & Replace tools followed by some minor edits to add our new parameter, or we can reach out to Copilot.

To rename the class reference parameter, we select the `validate_model` function again, open the inline Copilot chat, and provide the following prompt:

> Please use cls for the class parameter name

Find & Replace could have done this just as easily, and we might have had more confidence in the results, but changing the rest of the code to use the refactored function is a bit more complex. Let's try having Copilot helping us with the rest of the change.
```